### ISSUE TITLE
Missing documentation on how nontemporal stores interact with concurrency primitives

### ISSUE THREAD

**Issue by RalfJung** [llvm/llvm-project#64521]
The LLVM LangRef doesn't document how `!nontemporal` stores are intended to interact with concurrency primitives. The current interactions are extremely surprising, basically making `!nontemporal` stores even less ordered than "non-atomic" stores:
```
Thread A:
  store i32 %v, ptr %p, !nontemporal !13
  fence release
  // set some global flag (relaxed write)

Thread B:
  // wait till global flag is set (relaxed read)
  fence acquire
  %_0 = load i32, ptr %p, !noundef !11
```

According to all the usual concurrency rules, that last load must see the store. However, the way LLVM compiles this program on x86, it has a data race: the fences become NOPs and the relaxed accesses become regular MOV, so we end up with `MOVNT; MOV` in thread A, which the CPU is allowed to reorder -- meaning that thread B might see the flag write but then fail to see the data store!

In other words, `!nontemporal` violates TSO, but the compilation scheme LLVM (and everyone else) uses for release/acquire synchronization relies on TSO. Together this leads to rather unpredictable semantics.

**Comment by workingjubilee** [comment]
To be clear, this is slightly more than _just_ a documentation bug, as
- `!nontemporal` is a cross-ISA i.e. LLVM language annotation
- `atomic store ... release` or `fence release` are lowered to nops on x86 targets, due to assuming the ISA implements TSO for all store operations, instead of a highly modified interpretation that allows selectively weakening store-store ordering in the presence of the MOVNT* family of instructions
  - these TSO-weakening instructions can be selected during lowering of `store ... !temporal`
  - in the presence of weakened TSO, atomic synchronization instead requires SFENCE
- the only way of forcing a cross-ISA store-store synchronization that would get lowered correctly is the use of a much stronger ordering, like `fence seq_cst`!
- the only way to resolve this as a frontend without incurring the wastefully stronger fence is to use architecture-specific intrinsics like `llvm.x86.sse.sfence` or assembly, neither of which are resilient when targeting multiple architectures, as other ISAs have added nontemporal stores also (fortunately, they are ones which already had weaker atomic semantics, so `fence release` should still work... so far)
- that is, this has to be at least one of:
  - unsoundness
  - a missed optimization
  - ergonomics that makes simply writing machine-specific assembly look good

And Ralf's primary concern is about load and store reorderings due to optimizations not being mindful of these weaker-than-TSO orderings described by MOVNTDQ, MOVNTI, and so on, and the need for SFENCE to not be moved in some cases. It's not even clear how much we can rely on inline assembly (when emitted as separate statements or present in separate functions) not being reordered or elided by such compiler optimizations.

Thus, Ralf saying, "The LLVM LangRef doesn't document how so-and-so interacts with concurrency..." might, because he is a concurrency expert yet does not feel he knows how to resolve the concurrency problems here, be rendered in a more conversational dialect of English more like, "So, how fucked are we? Because this situation seems fucked up beyond all recognition."

Ideally, there would be a way to force `store !nontemporal` to come with emitting a metadata node of some kind that indicates the need for SFENCE. This could then be optimized to, upon encountering multiple `store !nontemporal` in close succession (say, after inlining), only emitting one actual SFENCE, until it reaches a `fence release`, `atomic store ... release`, or anything that would emit a much stronger barrier instead. It would also be eligible for usage in something analogous to the existing VZEROUPPER-insertion pass, to prevent returning to safe code without first reaching SFENCE.

**Comment by RalfJung** [comment]
Yeah, the current status seems to be that a `store !nontemporal` *not* followed by an sfence is just full UB, since release write and fences fail to have their documented behavior. If a `store !nontemporal` *is* eventually followed by an sfence, the machine is still in a strange state until that happens (namely a state where release operations don't behave as documented), and it is unclear to me if that can cause other issues (e.g. due to the compiler inserting release operations there -- obviously LLVM optimizations can generally assume that `release` always works as documented.

> other ISAs have added nontemporal stores also (fortunately, they are ones which already had weaker atomic semantics, so fence release should still work... so far)

I haven't looked into that at all -- how "strange" do nontemporal stores behave on those ISAs? Are they somehow weaker than regular non-atomic accesses, or is it truly just a hint to the CPU to not keep this cacheline cached?

**Comment by RalfJung** [comment]
I also haven't looked into nontemporal loads at all, and whether they can cause any havoc.

**Comment by workingjubilee** [comment]
re: nontemporal stores on AArch64:
- no special caveats, a "true hint"
- weakly ordered stores, so DMB ("Data Memory Barrier") is still needed to sort things out
- a note that shouldn't matter: there are slightly wider "access bands" the CPU is allowed to speculatively examine, these are always aligned though so we don't need to worry about e.g. hitting the next page, and it shouldn't be user-visible

re: nontemporal loads on AArch64:
- code can observe reorderings if the nontemporal loads interact with address dependencies ("load address, then load from that address")
- DMB still sorts things out as usual

**Comment by chorman0773** [comment]
Wasn't it noted that it wasn't just dependencies that don't interact properly, but a stale value of the address register could be used?

**Comment by jyknight** [comment]
There is no point in having `!nontemporal` in LLVM IR if it can't actually permit the use of the hardware instructions. Either the spec needs to be clarified to clearly allow this behavior, or the feature should just be removed. I would suggest the former.

Looking back, it seems this was discussed previously a decade ago, https://lists.llvm.org/pipermail/llvm-dev/2016-January/093912.html and then again 6 years ago, https://lists.llvm.org/pipermail/llvm-dev/2020-April/141278.html

BTW, the choice not to use the instructions on ARM seems to have come from a discussion on https://reviews.llvm.org/D12785?id=34511#inline-103735, which, as far as I can tell, seems to have arisen from a mistake in the ARM "Programmers guide", where they misinterpreted the actual architecture manual, to imply that the processor might actually to _execute_ a load of an address into a register before the use of the address. This is not in fact the case -- it only relaxes the _memory ordering_ guarantees, not the instruction register-dependency ordering. The same confusion was raised (and resolved) in the thread in https://lists.infradead.org/pipermail/linux-arm-kernel/2016-September/458253.html

**Comment by jyknight** [comment]
I believe the rules we need are:
- In a single-threaded execution, !nontemporal doesn't change the program semantics.
- However, a !nontemporal load/store _is not_ ordered by cross-thread synchronizes-with edges, unless there is a nontemporal fence between the non-temporal load/store and the instruction providing the synchronizes-with edge.
- Unfortunately, non-temporal fences are not available in LLVM IR "yet" (still, a decade after the initial proposal), so you must provide such fences in an architecture-specific manner for now (e.g. inline-asm).
